javascript/ql/src/semmle/javascript/security/dataflow/PrototypePollutingAssignment.qll

```diff
@@ -0,0 +1,205 @@
+  }
+
+  private DataFlow::SourceNode prototypeLessObject(DataFlow::TypeTracker t) {
+    t.start() and
+    // We assume the argument to Object.create is not Object.prototype, since most
+    // users wouldn't bother to call Object.create in that case.
+    result = DataFlow::globalVarRef("Object").getAMemberCall("create")
+    or
+    // Allow use of AdditionalFlowSteps and AdditionalTaintSteps to track a bit further
+    exists(DataFlow::Node mid |
+      prototypeLessObject(t.continue()).flowsTo(mid) and
+      any(DataFlow::AdditionalFlowStep s).step(mid, result)
+    )
+    or
+    exists(DataFlow::TypeTracker t2 | result = prototypeLessObject(t2).track(t2, t))
+  }
+
+  /** Holds if `Object.prototype` has a member named `prop`. */
+  private predicate isPropertyPresentOnObjectPrototype(string prop) {
+    exists(ExternalInstanceMemberDecl decl |
+      decl.getBaseName() = "Object" and
+      decl.getName() = prop
+    )
+  }
+
+  /** A check of form `e.prop` where `prop` is not present on `Object.prototype`. */
+  private class PropertyPresenceCheck extends TaintTracking::LabeledSanitizerGuardNode,
+    DataFlow::ValueNode {
+    override PropAccess astNode;
+
+    PropertyPresenceCheck() {
+      astNode = any(ConditionGuardNode c).getTest() and // restrict size of charpred
+      not isPropertyPresentOnObjectPrototype(astNode.getPropertyName())
+    }
+
+    override predicate sanitizes(boolean outcome, Expr e, DataFlow::FlowLabel label) {
+      e = astNode.getBase() and
+      outcome = true and
+      label instanceof ObjectPrototype
+    }
+  }
+
+  /** A check of form `"prop" in e` where `prop` is not present on `Object.prototype`. */
+  private class InExprCheck extends TaintTracking::LabeledSanitizerGuardNode, DataFlow::ValueNode {
+    override InExpr astNode;
+
+    InExprCheck() {
+      not isPropertyPresentOnObjectPrototype(astNode.getLeftOperand().getStringValue())
+    }
+
+    override predicate sanitizes(boolean outcome, Expr e, DataFlow::FlowLabel label) {
+      e = astNode.getRightOperand() and
+      outcome = true and
+      label instanceof ObjectPrototype
+    }
+  }
+
+  /** A check of form `e instanceof X`, which is always false for `Object.prototype`. */
+  private class InstanceofCheck extends TaintTracking::LabeledSanitizerGuardNode,
+    DataFlow::ValueNode {
+    override InstanceofExpr astNode;
+
+    override predicate sanitizes(boolean outcome, Expr e, DataFlow::FlowLabel label) {
+      e = astNode.getLeftOperand() and
+      outcome = true and
+      label instanceof ObjectPrototype
+    }
+  }
+
+  /** A check of form `typeof e === "string"`. */
+  private class TypeofCheck extends TaintTracking::LabeledSanitizerGuardNode, DataFlow::ValueNode {
+    override EqualityTest astNode;
+    Expr operand;
+    string value;
+
+    TypeofCheck() {
+      astNode.getLeftOperand().(TypeofExpr).getOperand() = operand and
+      astNode.getRightOperand().getStringValue() = value
```

Comment:
How about checks like the below? 

```JavaScript
"string" === typeof x
```

Reply:
Done. Looks like I also forgot to check the polarity of the equality check, added that as well.